libpod/runtime_pod_linux.go

```diff
@@ -87,45 +87,58 @@ func (r *Runtime) NewPod(options ...PodCreateOption) (*Pod, error) {
 	if pod.config.UsePodCgroup {
 		logrus.Debugf("Got pod cgroup as %s", pod.state.CgroupPath)
 	}
+	if pod.HasInfraContainer() != pod.SharesNamespaces() {
+		return nil, errors.Errorf("Pods must have an infra container to share namespaces")
+	}
 
 	if err := r.state.AddPod(pod); err != nil {
 		return nil, errors.Wrapf(err, "error adding pod to state")
 	}
 
+	if pod.HasInfraContainer() {
+		ctr, err := r.createInfraContainer(ctx, pod)
+		if err != nil {
+			// Tear down pod, as it is assumed a the pod will contain
+			// a pause container, and it does not.
+			if err2 := r.removePod(ctx, pod, true, true); err2 != nil {
+				logrus.Errorf("Error removing pod after pause container creation failure: %v", err2)
+			}
+			return nil, errors.Wrapf(err, "error adding Infra Container")
+		}
+		pod.state.InfraContainerID = ctr.ID()
+		if err := pod.save(); err != nil {
+			return nil, err
+		}
+	}
+
 	return pod, nil
 }
 
 func (r *Runtime) removePod(ctx context.Context, p *Pod, removeCtrs, force bool) error {
-	r.lock.Lock()
-	defer r.lock.Unlock()
-
-	if !r.valid {
-		return ErrRuntimeStopped
-	}
-
 	if !p.valid {
 		if ok, _ := r.state.HasPod(p.ID()); !ok {
-			// Pod was either already removed, or never existed to
-			// begin with
+			// Pod probably already removed
+			// Or was never in the runtime to begin with
 			return nil
 		}
 	}
 
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
-	// Force a pod update
-	if err := p.updatePod(); err != nil {
-		return err
-	}
-
 	ctrs, err := r.state.PodContainers(p)
 	if err != nil {
 		return err
 	}
 
 	numCtrs := len(ctrs)
 
+	// If the only container in the pod is the pause container, remove the pod and container unconditionally.
+	if err := p.updatePod(); err != nil {
```

Comment:
Shouldn't need a second updatePod() here - do one at the start of the function

Reply:
There's only one. I moved the one from above down here and I can't quite remember why.

Reply:
I'd move it back up to the top of the function, then